src/rgw/rgw_sal_motr.cc

```diff
@@ -2727,15 +2839,49 @@ int MotrAtomicWriter::complete(size_t accounted_size, const std::string& etag,
   string bucket_index_iname = "motr.rgw.bucket.index." + tenant_bkt_name;
   rc = store->do_idx_op_by_name(bucket_index_iname,
                                 M0_IC_PUT, obj.get_key().to_str(), bl);
-  if (rc == 0)
-    store->get_obj_meta_cache()->put(dpp, obj.get_key().to_str(), bl);
 
-  if (rc == 0 && !info.versioning_enabled())
-      // Delete old object data if exists.
-      old_obj.delete_mobj(dpp);
-
-  if (rc != 0)
+  if (rc != 0) {
     obj.delete_mobj(dpp);
+    return rc;
+  }
+
+  store->get_obj_meta_cache()->put(dpp, obj.get_key().to_str(), bl);
+
+  // update bucket stats in user stats index.
+  std::string user_stats_iname = "motr.rgw.user.stats." + owner.to_str();
+  bl.clear();
+  rgw_bucket_dir_header bkt_header;
+  rc = store->do_idx_op_by_name(user_stats_iname,
+                            M0_IC_GET, tenant_bkt_name, bl);
+  if (rc != 0) {
+    ldpp_dout(dpp, 20) << __func__ << ": Failed to get the bucket header."
+      << " bucket = " << tenant_bkt_name << ", ret = " << rc << dendl;
+    return rc;
+  }
+  
+  bufferlist::const_iterator bitr = bl.begin();
+  bkt_header.decode(bitr);
+  rgw_bucket_category_stats& bkt_stat = bkt_header.stats[RGWObjCategory::Main];
+  bkt_stat.total_size += total_data_size - old_obj.get_obj_size();
```

Comment:
This should be in if block  .. if not versioning_enabled

Reply:
It does not matter, if the old object does not exist, its size is defaulted to 0.

Reply:
got the mistake 👍🏼

Reply:
addressed @sachinpunadikar, tested with the versioned` object as well.
```
╰─➤  aws s3api list-object-versions --bucket mybkt-v1 --profile sumedh
{
    "Versions": [
        {
            "ETag": "\"da39978011e9e121b96e74bc4cf0c737\"",
            "Size": 142044,
            "StorageClass": "STANDARD",
            "Key": "obj1",
            "VersionId": "zWszE5bg845PS9aMMI3Py2R3TBr0mYs",
            "IsLatest": true,
            "LastModified": "2022-05-27T05:59:39.977Z",
            "Owner": {
                "DisplayName": "SumedhAK",
                "ID": "sumedh"
            }
        },
        {
            "ETag": "\"fb92234566bc9f01b853318711106da7\"",
            "Size": 1024,
            "StorageClass": "STANDARD",
            "Key": "obj1",
            "VersionId": "zWszEwvEVllzAOCv9UaipHoEK0Hjxgu",
            "IsLatest": false,
            "LastModified": "2022-05-27T05:56:15.036Z",
            "Owner": {
                "DisplayName": "SumedhAK",
                "ID": "sumedh"
            }
        },
        {
            "ETag": "\"9cce32b2829b5c73b70075cef7210458\"",
            "Size": 1048576,
            "StorageClass": "STANDARD",
            "Key": "obj1",
            "VersionId": "zWszF3WPjn5lcYzjq4dMnj7G6e6vzuk",
            "IsLatest": false,
            "LastModified": "2022-05-27T05:55:49.679Z",
            "Owner": {
                "DisplayName": "SumedhAK",
                "ID": "sumedh"
            }
        }
    ]
}
```

```
"usage": {
        "rgw.main": {
            "size": 1191644,
            "size_actual": 1191936,
            "size_utilized": 1191644,
            "size_kb": 1164,
            "size_kb_actual": 1164,
            "size_kb_utilized": 1164,
            "num_objects": 3
        }
    }
```

Reply:
This looks alright.